crypto/keyring/output.go

```diff
@@ -1,106 +1,73 @@
 package keyring
 
 import (
+	"github.com/cosmos/cosmos-sdk/codec"
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
+// TODO: Move this file to client/keys
+// Use protobuf interface marshaler rather then generic JSON
+
 // KeyOutput defines a structure wrapping around an Info object used for output
 // functionality.
 type KeyOutput struct {
-	Name      string                 `json:"name" yaml:"name"`
-	Type      string                 `json:"type" yaml:"type"`
-	Address   string                 `json:"address" yaml:"address"`
-	PubKey    string                 `json:"pubkey" yaml:"pubkey"`
-	Mnemonic  string                 `json:"mnemonic,omitempty" yaml:"mnemonic"`
-	Threshold uint                   `json:"threshold,omitempty" yaml:"threshold"`
-	PubKeys   []multisigPubKeyOutput `json:"pubkeys,omitempty" yaml:"pubkeys"`
+	Name     string `json:"name" yaml:"name"`
+	Type     string `json:"type" yaml:"type"`
+	Address  string `json:"address" yaml:"address"`
+	PubKey   string `json:"pubkey" yaml:"pubkey"`
+	Mnemonic string `json:"mnemonic,omitempty" yaml:"mnemonic"`
 }
 
 // NewKeyOutput creates a default KeyOutput instance without Mnemonic, Threshold and PubKeys
-func NewKeyOutput(name, keyType, address, pubkey string) KeyOutput {
+func NewKeyOutput(name string, keyType KeyType, a sdk.Address, pk cryptotypes.PubKey) (KeyOutput, error) { // nolint:interfacer
+	bz, err := codec.ProtoMarshalJSON(pk, nil)
+	if err != nil {
+		return KeyOutput{}, err
+	}
 	return KeyOutput{
 		Name:    name,
-		Type:    keyType,
-		Address: address,
-		PubKey:  pubkey,
-	}
-}
-
-type multisigPubKeyOutput struct {
-	Address string `json:"address" yaml:"address"`
-	PubKey  string `json:"pubkey" yaml:"pubkey"`
-	Weight  uint   `json:"weight" yaml:"weight"`
+		Type:    keyType.String(),
+		Address: a.String(),
+		PubKey:  string(bz),
```

Comment:
This will try to escape `"`s, e.g.

```json
{
  "name": "abc",
  "pub_key": "{\"@type\": \"...\"}"
}
```

I think for now it's fine, but ideally `KeyOutput` would be a proto message.

Reply:
Yes, but this is out of the scope. Main thing here was to change the PubKey type.